[inductor] Make autotune_process.py pass mypy

jkseppan opened the pull request:
`TensorMeta.from_irnodes` handles either a single `IRNode` or a tuple or list of them. I tried to express this with overloading, but because this file is in MYPYNOFOLLOW, the `IRNode` subclasses become `Any`, which causes the overloads to be overlapping.

This changes the type of the argument to `benchmark_in_sub_process` to the more specific `TritonTemplateCaller`, since that one has the `bmreq` member and existing docstrings indicate that only the triton template benchmark is handled.

The `rand_strided` call caused a mypy error because the default value for device was a string. This is fixed by adding type hints to `rand_strided` in `torch/_dynamo/testing.py`. Likewise, the return value of `PyCodeCache.load_by_key_path` can be inferred from the type hint on `PyCodeCache.cache`.

Fixes one part of #105230


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

jkseppan committed:
[inductor] Make autotune_process.py pass mypy

TensorMeta.from_irnodes handles either a single IRNode or a tuple or
list of them. I tried to express this with overloading, but because
this file is in MYPYNOFOLLOW, the IRNode subclasses become Any, which
causes the overloads to be overlapping.

This changes the type of the argument to benchmark_in_sub_process to
the more specific TritonTemplateCaller, since that one has the bmreq
member and existing docstrings indicate that only the triton template
benchmark is handled.

The rand_strided call caused a mypy error because the default value
for device was a string. This is fixed by adding type hints to
rand_strided in torch/_dynamo/testing.py. Likewise, the return value
of PyCodeCache.load_by_key_path can be inferred from the type hint on
PyCodeCache.cache.

See #105230

jkseppan commented:
@pytorchbot label "topic: not user facing"

Skylion007 commented:
Instead of fixing all these Optionals manually, we should just enable ruff rule RUF013

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -1,8 +1,14 @@
+from __future__ import annotations
```
I don't think this __future__ import annotations is necessary here, can see if it works without it?

jkseppan replied:
Not strictly necessary, but without it we need to use strings for several of the annotations.

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -201,14 +241,14 @@ def worker():
     """
     assert choice.bmreq is not None
     tuning_process.initialize()
-    assert tuning_process.valid()
```
This does not need to be transformed into a class method? All it needs is an isinstance assert so mypy can properly determine the type.

Skylion007 replied:
Is this because attrs of class methods are propagated but normal methods are not? Regardless, the type hints on valid_process are wrong.

jkseppan replied:
(this part of the diff got reverted)

jkseppan commented on torch/_inductor/autotune_process.py:
```diff
@@ -133,7 +169,7 @@ def to_tensor(self) -> torch.Tensor:
             self.sizes,
             self.strides,
-            device=self.device,
+            device=self.device.type,
```
This change is probably unnecessary, `rand_strided` just needs a type annotation for the `device` parameter. The default value is a string but it looks like a `torch.device` should work too.

jkseppan replied:
Annotating the arguments of `rand_strided` is clearly the better solution

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -50,21 +58,25 @@ def process_main(request_queue, response_queue):
+        return self.valid_process(self)
+
+    @classmethod
+    def valid_process(cls, ob: "TuningProcess") -> TypeGuard["ValidTuningProcess"]:
```
The return type here is wrong, it returns a boolean, right?

jkseppan replied:
Yes, it returns a boolean but TypeGuard is a special value from [PEP 647](https://peps.python.org/pep-0647/) that lets mypy know that the instance can be treated as another class where the Optional fields are not optional. This lets you avoid a bunch of type assertions when you have already checked that the fields are not None.

But now I realize that PEP 647 is only in Python 3.10 and up, so I suppose I'll have to find another way.

jkseppan replied:
I replaced this with just casting the fields to non-Optional values - not too much line noise, and probably simpler to understand than the TypeGuard solution anyway.

Skylion007 replied:
@jkseppan What about the solution I just proposed? Or is that also too much line noise?

jkseppan replied:
I implemented your suggestion. It has the advantage of being very explicit, and like you said, it lets the type inference do its thing and doesn't hide other type errors. Ideally mypy would be able to look inside the `valid()` call, but as of now it doesn't do that.

jkseppan commented:
Rebased on top of main to get the linter upgrades

Skylion007 commented:
Why are the casts necessary over just asserting the attrs is not None?

jkseppan commented:
> Why are the casts necessary over just asserting the attrs is not None?

The casts are no-ops at runtime, they are just there for mypy. The asserts are in another method, but mypy doesn't look into it recursively so they would have any effect in the calling function. This is why I first made this method return a TypeGuard value, which is the mechanism for telling type checkers that a function can be used to check that its argument adheres to a stricter type. But that was many more lines, so I can see how it might have been seen as confusing.

kit1980 commented:
@jkseppan please resolve the conflict

jkseppan commented:
Rebased - it seems that some type annotations were added elsewhere.

The `@overload` annotations were causing weird errors, which I finally realized were because this file is in MYPYNOFOLLOW. The `ir.Node` subclasses were treated the same as `Any`, which overlaps with sequence types, so it's not really possible to express "we return a list for a sequence input and a single value for a single input" at least with overloading. This could be revisited if the follow-imports setting can eventually be enabled.

jkseppan committed:
Quote the Queue[Any] annotations

Mypy requires a type parameter but then some tests fail with `TypeError:
'type' object is not subscriptable`.

jkseppan commented:
@Skylion007 Did I address your concerns sufficiently? If I read the contributing guidelines correctly, I could now ask the bot to merge since there is one approving review, but I'm a newbie here and don't want to merge over anyone's objections.

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -81,7 +85,7 @@ def initialize(self) -> None:
                 self.response_queue,
             ),
         )
-        self.process.start()
```
the proper way to fix this 
```
process = self.process
assert process is not None
process.start()
```

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -93,19 +97,22 @@ def initialize(self) -> None:
         # wait for the initialization to be done
-        self.request_queue.put(Ping())
-        resp = self.response_queue.get()
+        cast("Queue[Any]", self.request_queue).put(Ping())
+        resp = cast("Queue[Any]", self.response_queue).get()
```
the proper way to fix this 
```suggestion
        process = self.process
        assert process is not None
        process.start()
```

Skylion007 reviewed (changes requested):
There is a little annoying but way safer to handle these casts.

Skylion007 commented:
> > Why are the casts necessary over just asserting the attrs is not None?
> 
> The casts are no-ops at runtime, they are just there for mypy. The asserts are in another method, but mypy doesn't look into it recursively so they would have any effect in the calling function. This is why I first made this method return a TypeGuard value, which is the mechanism for telling type checkers that a function can be used to check that its argument adheres to a stricter type. But that was many more lines, so I can see how it might have been seen as confusing.

They are noops at runtime but they can mask other typing issues and do have some minor perf penalty. I would advise doing what I suggested to make mypy happy (move it into a temp variable), but it's fine either way I suppose.

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -93,19 +97,22 @@ def initialize(self) -> None:
-            self.request_queue.put(None)
-            self.process.join()
+            cast("Queue[Any]", self.request_queue).put(None)
+            cast(BaseProcess, self.process).join()
```
the proper way to fix this 
```suggestion
            process = self.process
            assert process is not None
            process.join()
```

jkseppan replied:
Sure, thanks; I handled the other casts similarly.

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -31,13 +34,14 @@ class Pong:
-    request_queue: multiprocessing.Queue = None
-    response_queue: multiprocessing.Queue = None
+    process: Optional[BaseProcess] = None
+    request_queue: "Optional[Queue[Any]]" = None
+    response_queue: "Optional[Queue[Any]]" = None
```
```suggestion
    request_queue: Optional["Queue[Any]"] = None
    response_queue: Optional["Queue[Any]"] = None
```

jkseppan committed:
Quote only Queue[Any]

Co-authored-by: Aaron Gokaslan <aaronGokaslan@gmail.com>

jkseppan committed:
Avoid casting

Put the values in local variables and check that they are not None,
allowing type inference to take care of the exact type.

Co-authored-by: Aaron Gokaslan <aaronGokaslan@gmail.com>

jkseppan commented:
@pytorchmergebot merge

jansel commented:
@pytorchmergebot merge
